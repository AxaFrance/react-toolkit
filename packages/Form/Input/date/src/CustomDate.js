import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import { parse, format as dateFormat } from 'date-fns';
import { Input, withInput } from '@axa-fr/react-toolkit-form-core';

const CustomDate = (props) => {
  const {
    className,
    classModifier,
    componentClassName,
    id,
    name,
    viewValue,
    value,
    locale,
    format,
    readOnly,
    disabled,
    onChangeRaw,
    onChange,
    ...otherProps
  } = props;

  const [dateFnsLocale, setDateFnsLocale] = useState(null);

  useEffect(() => {
    if (locale) {
      import(`date-fns/locale/${locale}/index.js`).then((loc) => {
        setDateFnsLocale(loc);
      });
    }
  }, [locale]);

  let currentViewValue = '';
  if (value) {
    currentViewValue = dateFormat(value, format);
  } else if (viewValue != null && viewValue != undefined) {
    currentViewValue = viewValue;
  }

  return (
    dateFnsLocale && (
      <DatePicker
        id={id}
        selected={value}
        name={name}
        onChange={onChange}
        onChangeRaw={onChangeRaw}
        value={currentViewValue}
        locale={dateFnsLocale}
        className={componentClassName}
        readOnly={readOnly}
        disabled={disabled}
        {...otherProps}
      />
    )
  );
};

const propTypes = {
  value: PropTypes.object,
  viewValue: PropTypes.string,
  locale: PropTypes.string,
  format: PropTypes.string,
};
const defaultClassName = 'af-datepicker';
const defaultProps = {
  value: null,
  viewValue: null,
  locale: 'fr',
  format: 'dd/MM/yyyy',
  className: defaultClassName,
  fixedHeight: true,
  showMonthDropdown: true,
  showYearDropdown: true,
  autoFocus: false,
  todayButton: "Aujourd'hui",
  popperPlacement: 'right-start',
  yearDropdownItemNumber: 6,
};

const handlers = {
  onChange: ({ id, name, onChange, format }) => (date) => {
    const viewValue = date ? dateFormat(date, format) : '';
    onChange({
      value: date,
      viewValue,
      name,
      id,
    });
  },
  onChangeRaw: ({ format, id, name, onChange }) => (event) => {
    const dateString = event.target.value;
    if (dateString) {
      const date = parse(dateString, format, new Date());
      onChange({
        value: date,
        viewValue: dateString,
        name,
        id,
      });
    }
  },
};

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  handlers
)(CustomDate);

EnhancedComponent.Clone = Input.Clone;
EnhancedComponent.displayName = CustomDate.name;

export default EnhancedComponent;
