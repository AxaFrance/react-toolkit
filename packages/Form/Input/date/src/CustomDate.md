## CustomDate

---

### Get started

First to use, you have to import the component.

```js
import { CustomDate } from 'af-toolkit-react';
import '@axa-fr/react-toolkit-form-input-date-phone/dist/datepicker.scss';
```

The component reuses the following library:

- react-datepicker: https://reactdatepicker.com

The date is a special component. It actually has 2 important props:

- viewValue: The value of the text field which is a string
- value: The value of the date in momentjs format

This allows you to have control of what happens and to respond to all use cases. If you are an individual, you can use the "viewValue" or "value" property. In any case if "viewValue" is not a valid date in the current language, the value of "value" returned will always be "null".


### Supported Languages

The Date input allow to use a differents languages : Spanish(es-ES), English (en-GB or en-US), French (fr-FR) and German (de-DE). By default french is used, you can specify the langage in the locale property using the . If you want a new language, please create a [new Issue](https://github.com/AxaGuilDEv/react-toolkit/issues)

### Format

Since the Version 2.0, The DateInput uses [Date-Fns](https://date-fns.org/) and return a Javascript Date object. Moment is not required anymore. You can specify the string format for viewValue in the format property. The default format is "dd/MM/YYYY" you can find the list on the [Date-fns](https://date-fns.org/v2.0.0-alpha.18/docs/format) Website

For more information about style component, you can see the [Form Date Style on Toolkit-core ](http://toolkit-intranet-axa.azurewebsites.net/#/form)
