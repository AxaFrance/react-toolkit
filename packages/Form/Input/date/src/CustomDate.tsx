import React, { ComponentPropsWithRef, forwardRef } from 'react';
import { ClassManager } from '@axa-fr/react-toolkit-form-core';

type Props = Omit<ComponentPropsWithRef<'input'>, 'value' | 'onChange'> & {
  classModifier?: string;
  onChange?: (date: Date) => void;
  locale?: string | string[];
  value: Date;
};

const Date = forwardRef<HTMLInputElement, Props>(
  (
    { className, classModifier, value, locale, onChange, ...otherProps },
    ref
  ) => {
    const monthFormatted = `0${value.getMonth() + 1}`.slice(-2);
    const dayFormatted = `0${value.getDate()}`.slice(-2);
    const currentValue = `${value.getFullYear()}-${monthFormatted}-${dayFormatted}`;

    const componentClassName = ClassManager.getComponentClassName(
      className,
      classModifier,
      'af-form__input-date'
    );

    return (
      <input
        className={componentClassName}
        type="date"
        value={currentValue}
        onChange={(e) => {
          const newValue = e.currentTarget.valueAsDate;
          onChange && newValue && onChange(newValue);
        }}
        ref={ref}
        {...otherProps}
      />
    );
  }
);

export default Date;
