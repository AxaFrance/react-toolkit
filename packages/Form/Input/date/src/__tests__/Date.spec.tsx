import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import CustomDate from '../Date';

describe('Date component', () => {
  it('renders with initial value', () => {
    const initialValue = new Date('2024-03-04');
    const { asFragment, getByDisplayValue } = render(
      <CustomDate value={initialValue} />
    );

    getByDisplayValue('2024-03-04');

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onChange handler when input value changes', () => {
    const onChangeMock = jest.fn();
    const initialValue = new Date('2024-03-04');
    const { getByLabelText, asFragment } = render(
      <>
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        <label htmlFor="date">Date:</label>
        <CustomDate
          onChange={onChangeMock}
          name="date"
          id="date"
          value={initialValue}
        />
      </>
    );
    const inputElement = getByLabelText('Date:') as HTMLInputElement;

    userEvent.type(inputElement, '2024-03-05');
    expect(inputElement.value).toBe('2024-03-05');

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onChange handler when input value changes although target value is null', () => {
    const initialValue = new Date('2024-03-04');
    const onChangeMock = jest.fn();
    const { getByLabelText, asFragment } = render(
      <>
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        <label htmlFor="date">Date:</label>
        <CustomDate
          onChange={onChangeMock}
          name="date"
          id="date"
          value={initialValue}
        />
      </>
    );
    const inputElement = getByLabelText('Date:') as HTMLInputElement;

    userEvent.clear(inputElement);

    expect(inputElement.value).toBe('');

    expect(asFragment()).toMatchSnapshot();
  });

  it('applies class modifier', () => {
    const { container, asFragment } = render(
      <CustomDate classModifier="required" />
    );
    container.querySelector('.af-form__input-date--required');

    expect(asFragment()).toMatchSnapshot();
  });
});
