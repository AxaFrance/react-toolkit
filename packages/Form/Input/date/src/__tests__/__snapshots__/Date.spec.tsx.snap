// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date component applies class modifier 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-date af-form__input-date--required"
    required=""
    type="date"
    value=""
  />
</DocumentFragment>
`;

exports[`Date component calls onChange handler when input value changes 1`] = `
<DocumentFragment>
  <label
    for="date"
  >
    Date:
  </label>
  <input
    class="af-form__input-date"
    id="date"
    name="date"
    type="date"
    value="2024-03-04"
  />
</DocumentFragment>
`;

exports[`Date component calls onChange handler when input value changes although target value is null 1`] = `
<DocumentFragment>
  <label
    for="date"
  >
    Date:
  </label>
  <input
    class="af-form__input-date"
    id="date"
    name="date"
    type="date"
    value="2024-03-04"
  />
</DocumentFragment>
`;

exports[`Date component renders with initial value 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-date"
    type="date"
    value="2024-03-04"
  />
</DocumentFragment>
`;
